extensions/scarb-prove/src/main.rs

```diff
@@ -0,0 +1,185 @@
+
+    /// ID of `scarb execute` *standard* output for given package, for which to generate proof.
+    #[arg(long)]
+    execution_id: Option<usize>,
+
+    /// Execute the program before proving.
+    #[arg(
+        long,
+        default_value_t = false,
+        conflicts_with = "execution_id",
+        required_unless_present = "execution_id"
+    )]
+    execute: bool,
+
+    #[command(flatten)]
+    execute_args: ExecutionArgs,
+
+    #[command(flatten)]
+    prover: ProverArgs,
+
+    /// Logging verbosity.
+    #[command(flatten)]
+    pub verbose: VerbositySpec,
+}
+
+#[derive(Parser, Clone, Debug)]
+struct ProverArgs {
+    /// Track relations during proving.
+    #[arg(long, default_value = "false")]
+    track_relations: bool,
+
+    /// Display components during proving.
+    #[arg(long, default_value = "false")]
+    display_components: bool,
+}
+
+fn main() -> ExitCode {
+    let args = Args::parse();
+    let ui = Ui::new(args.verbose.clone().into(), OutputFormat::Text);
+
+    match main_inner(args, ui.clone()) {
+        Ok(()) => ExitCode::SUCCESS,
+        Err(error) => {
+            ui.error(format!("{error:#}"));
+            ExitCode::FAILURE
+        }
+    }
+}
+
+fn main_inner(args: Args, ui: Ui) -> Result<()> {
+    let scarb_target_dir = Utf8PathBuf::from(env::var("SCARB_TARGET_DIR")?);
+
+    ui.warn("soundness of proof is not yet guaranteed by Stwo, use at your own risk");
+
+    let metadata = MetadataCommand::new().inherit_stderr().exec()?;
+    let package = args.packages_filter.match_one(&metadata)?;
+
+    let execution_id = match args.execution_id {
+        Some(id) => id,
+        None => {
+            ensure!(args.execute);
+            run_execute(&args, &ui)?
+        }
+    };
+    ui.print(Status::new("Proving", &package.name));
+
+    let (pub_input_path, priv_input_path, proof_path) =
+        resolve_paths_from_package(&scarb_target_dir, &package.name, execution_id)?;
+
+    let prover_input = adapt_vm_output(
+        pub_input_path.as_std_path(),
+        priv_input_path.as_std_path(),
+        false,
+    )
+    .context("failed to adapt VM output")?;
+
+    let config = ProverConfig::builder()
+        .track_relations(args.prover.track_relations)
+        .display_components(args.prover.display_components)
+        .build();
+
+    let proof = prove_cairo::<Blake2sMerkleChannel>(prover_input, config)
+        .context("failed to generate proof")?;
+
+    ui.print(Status::new(
+        "Saving proof to:",
+        &display_path(&scarb_target_dir, &proof_path),
+    ));
+
+    fs::write(proof_path.as_std_path(), serde_json::to_string(&proof)?)?;
+
+    Ok(())
+}
+
+fn run_execute(args: &Args, ui: &Ui) -> Result<usize> {
+    let args = ExecuteArgs {
+        packages_filter: args.packages_filter.clone(),
+        execution: args.execute_args.clone(),
+        verbose: args.verbose.clone(),
+    };
+    let execution_id = scarb_execute::main_inner(args, ui.clone())?;
```

Comment:
I think we should pass package metadata here, instead of calling ScarbMetadata resolution a second time. The function name could also be more descriptive.